Objects/typeobject.c

```diff
@@ -488,14 +488,16 @@ PyType_Modified(PyTypeObject *type)
 
     if (type->tp_watched) {
         PyInterpreterState *interp = _PyInterpreterState_GET();
-        char bits = type->tp_watched;
-        for (int i = 0; i < TYPE_MAX_WATCHERS; i++) {
+        int bits = type->tp_watched;
+        int i = 0;
+        while(bits && i < TYPE_MAX_WATCHERS) {
```

Comment:
You don't need `i` anymore, right? If no bits are set, there is no work to do, so the loop exits; if some bits are set, there is still work to do, so the loop continues.

Reply:
@erlend-aasland We still need `i` because it is the index into `interp->type_watchers` array.

It is true that we technically don't need to check `i < TYPE_MAX_WATCHERS` here in the while condition, because the source of `bits` (`type->tp_watched`) is only 8 bits wide, so it should not be possible for any bit above bit 7 in `bits` to ever be set. But given that a too-large `i` could result in an out-of-bounds access on `interp->type_watchers`, I'm slightly uncomfortable relying fully on that subtlety. And I also don't think that type modifications should generally be in a super hot path.

Reply:
Well, `interp->type_watchers` is already guarded by `validate_watcher_id`, so this part of the code could just use an assert. Anyway, as you say, this should not be a super hot path, so it probably does not matter much :)

Reply:
You're right, given that `validate_watcher_id` wouldn't let any too-high bit be set in `tp_watched`, this check could probably just be an assert. I'll consider changing it if back in this area again in future.